File: DSA/Sort/quickselect.py
import random
import unittest


def swpFunc(arr):
    def swp(i, j): arr[i], arr[j] = arr[j], arr[i]
    return swp


def partitonTwoWayFunction(arr):
    swp = swpFunc(arr)

    def partition(start, end):
        assert start <= end
        pivot = start  # random.randint(start, end)
        pv = arr[pivot]
        swp(start, pivot)
        le = start
        gt = end
        while le < gt:
            if arr[gt] <= pv:
                le += 1
                swp(le, gt)
            else:
                gt -= 1

        swp(start, le)
        return le
    return partition


def quickSelect(arr, kth):
    partiton = partitonTwoWayFunction(arr)
    start, end = 0, len(arr) - 1
    while True:
        p = partiton(start, end)
        if p < kth:
            start = p + 1
        elif p > kth:
            end = p - 1
        else:
            return


def findKthLargest(arr, k):
    n = len(arr)
    quickSelect(arr, n - k)
    return arr[-k]


def findKthSmallest(arr, k):
    quickSelect(arr, k - 1)
    return arr[k]


# class TestQuickSelect(unittest.TestCase):
#     def test_findKthLargest(self):
if __name__ == '__main__':
    arr = [3, 2, 3, 1, 2, 4, 5, 5, 6]
    # arr = [1, 1]

    k = 2
    test = arr.copy()
    kthLargest = findKthLargest(test, k)
    print(kthLargest)
    assert kthLargest == 5

File: DSA/Sort/quickselect.cpp
#include <iostream>
#include <vector>
#include <random>
using namespace std;

typedef vector<int> vi;

void swp(vi &arr, int i, int j)
{
    int tmp = arr[j];
    arr[j] = arr[i];
    arr[i] = tmp;
}

int partition(vi &arr, int start, int end)
{
    int pivot = start + (rand() % (end - start + 1));
    int pv = arr[pivot];
    int le = start, gt = end;

    while (le < gt)
    {
        if (arr[gt] > pv)
        {
            gt -= 1;
        }
        else
        {
            le += 1;
            swp(arr, le, gt);
        }
    }
    swp(arr, le, start);
    return le;
}

void quickSelect(vi &arr, int k)
{

    int start = 0, end = arr.size() - 1;
    int p = 0;
    while (true)
    {
        p = partition(arr, start, end);
        if (k < p)
            end = p - 1;

        else if (p < k)
            start = p + 1;
        else
            return;
    }
}

int findKthLargest(vi arr, int k)
{

    quickSelect(arr, arr.size() - k);
    return arr[arr.size() - k];
}

int main()
{
    vi arr = {3, 3, 2, 1, 5, 6, 4};
    int kth = findKthLargest(arr, 4);
    printf("Kth largest element %d\n", kth);

    return 0;
}